### PR TITLE
Sidebar item selection fix

### PR DESCRIPTION
**Pull request by ronso0**
Fixes some issues when navigating the sidebar with Up/Down keys, `[Library],Move...`, `[Playlist],Select...` controls (emulated key presses)

**part1**
fixes https://bugs.launchpad.net/mixxx/+bug/1939082

Keys can NOT select the pevious/next playlist (Playlists, History) but make the selection jump to Tracks (and sometimes apply the key press after that).

Looks like #2996 added the redundant and failing code (git bisect points to a8a93f02e44751d42e9673d622dcd0233c024feb), ~~though I didn't trace what exactly causes the fail~~ (redundant sidebar item activation with invalid index) 

En passant I added some debug assertions.

**part 2**
After creating a playlist/crate Up/Down keys would make the selection jump to the Tracks.
Fixed by adding `setCurrentIndex()` to  `WLibrarySidebar::selectIndex` and `WLibrarySidebar::selectChildIndex`.
**edit** also prevents the jump when moving away from active external features' root item (Serato, Rekordbox ...)

**part 3**
Remove a crate and the selection jumps to the Crates root item.
Fix: Like with playlists, before deleting a crate, store its neighbour's Id to somewhat restore the selection afterwards, or rather scroll position.

### CODE REVIEW

**Comment by ronso0** [comment]
>  though I didn't trace what exactly causes the fail

Any `select prev/next sidebar item` command is either an actual or emulated 
That is captured by the sidebar model's clickTimer, which then invokes `pFeature->activateChild(index);`, so the additonal `featureSelect` and `m_pSidebarWidget->selectChildIndex(index)` stuff is entirely superflous (and wrong since it is invoked with an invalid index).

I'm puzzled that no one -incl me- noticed the bug since the History tweaks were merged. Obviously we weren't using controllers for navigating the playlists or history.

File: src/library/trackset/crate/cratefeature.cpp [hunk 1]
```diff
@@ -451,9 +450,7 @@ void CrateFeature::slotDuplicateCrate() {
         CrateId crateId =
```
Comment (uklotzde) [inline src/library/trackset/crate/cratefeature.cpp:450]: duplicateCrate() may return an invalid id if the user aborted the action. In this case the last crate should remain activated.

File: src/library/trackset/crate/cratefeature.cpp [hunk 2]
```diff
@@ -382,9 +383,7 @@ void CrateFeature::slotCreateCrate() {
     CrateId crateId =
             CrateFeatureHelper(m_pTrackCollection, m_pConfig)
                     .createEmptyCrate();
-    if (crateId.isValid()) {
-        activateCrate(crateId);
-    }
+    activateCrate(crateId);
```
Comment (uklotzde) [inline src/library/trackset/crate/cratefeature.cpp:386]: Same here, the creation could be aborted by the user. Please revert those unrelated changes.

File: src/library/trackset/baseplaylistfeature.cpp [hunk 3]
```diff
@@ -179,17 +179,9 @@ void BasePlaylistFeature::activateChild(const QModelIndex& index) {
     if (playlistId == kInvalidPlaylistId) {
         return;
     }
-
     m_pPlaylistTableModel->setTableModel(playlistId);
     emit showTrackModel(m_pPlaylistTableModel);
     emit enableCoverArtDisplay(true);
-    // Update selection
```
Comment (uklotzde) [inline src/library/trackset/baseplaylistfeature.cpp:186]: `emit featureSelect()` is invoked in 2.3 and everything works as expected. What was the reason for not emitting this signal?

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/library/trackset/crate/cratefeature.cpp (hunk 1 above)
Comment (ronso0) [reply src/library/trackset/crate/cratefeature.cpp:450]: activatCrate() has it's own check, thus I removed it here. but I don't understand why double-check because failure has no consequences inside this method.
https://github.com/mixxxdj/mixxx/blob/0d58e4ee2a2dc7ae3626fda5e84b07517d57c5c6/src/library/trackset/crate/cratefeature.cpp#L294-L298

Unrelated, yes, so I'll revert.

Comment (uklotzde) [reply src/library/trackset/crate/cratefeature.cpp:450]: The check in `CrateFeature::activateCrate()` is a debug assertion that we don't want to hit.

I am not an advocate of VERIFY_OR_DEBUG_ASSERT, which weakens the purpose of debug assertions. Even if the precondition is checked twice it is still an assertion. If `CrateFeature::activateCrate()` should accept invalid crate ids then we need to replace this assertion with a regular conditional.

File: src/library/trackset/baseplaylistfeature.cpp (hunk 3 above)
Comment (ronso0) [reply src/library/trackset/baseplaylistfeature.cpp:186]: > `emit featureSelect()` is invoked in 2.3 and everything works as expected.

you're confusing it with activatePlaylist()?
https://github.com/mixxxdj/mixxx/blob/05016f8c223b836977e409862b3e6d6feceff41d/src/library/baseplaylistfeature.cpp#L155-L163
Also `featureSelect()` is redundant here, and it's currently only updating the selection NOT the currentIndex(), unlike when the selection is changed by a (real or emulated) key press inside `WLibrarySidebar::keyPressEvent`.
> it's currently only updating the selection NOT the currentIndex()

I fixed that with the last commit.

File: src/widget/wlibrarysidebar.cpp [hunk 4]
```diff
@@ -232,6 +237,9 @@ void WLibrarySidebar::selectChildIndex(const QModelIndex& index, bool selectItem
         return;
     }
     QModelIndex translated = sidebarModel->translateChildIndex(index);
+    if (!translated.isValid()) {
```
Comment (uklotzde) [inline src/widget/wlibrarysidebar.cpp:240]: Is this expected or shall we use VERIFY_OR_DEBUG_ASSERT here?

**Comment by ronso0** [comment]
I fixed another bug: after creating, deleting or renaming a playlist/crate the selection also jumps unexpedtedly. This is due to missing setCurrentIndex() in `WLibrarySidebar::selectIndex` and `WLibrarySidebar::selectChildIndex`. I suppose this was the root cause for redundant `featureSelect()` to fail.

File: src/widget/wlibrarysidebar.cpp (hunk 4 above)
Comment (ronso0) [reply src/widget/wlibrarysidebar.cpp:240]: yes, could happen, for example from ... though I don't know how/when concurrent playlist manipulation would happen
https://github.com/mixxxdj/mixxx/blob/e5a2b129f15e637d53523d865f1f11647847059d/src/library/trackset/baseplaylistfeature.cpp#L419-L421

**Review by uklotzde — APPROVED** [review]
After a manual test we can merge.

**Comment by ronso0** [comment]
I'm adding another selection fix right now, see the TODO
https://github.com/mixxxdj/mixxx/pull/4193/files#diff-784192df9ca0c20864aeee323d8c9f51193221742b69a1faf71b0ecf0bc5559bR398-R399

If you don't mind Iet's include that,too

**Comment by ronso0** [comment]
@uklotzde Please take a look at the last one.
The sidebar selection should now be usable after any crate or playlist operation.

**Comment by Swiftb0y** [comment]
I can confirm that part 1 is fixed. I had trouble reproducing part 2 & 3 on main so I can't comment on those.

**Comment by ronso0** [comment]
@Swiftb0y I updated the 1st post with steps how to reproduce. Both keyboard and Library/Playlist controls are affected, except for part3 where the selection always jumps.

**Comment by Swiftb0y** [comment]
I'm sorry, my controller seems to be broken right now so I have no way to test this anymore.

**Comment by ronso0** [comment]
Library, Move.../Playlist,Select... controls are just emulated key presses in the sidebar widget, so a keyboard is sufficient for testing. that's also how I noticed the bugs in the first place.

**Comment by Swiftb0y** [comment]
Thanks, I thought these issues were only present when navigating via a controller because I had trouble reproducing part 2 and 3.

**Review by Swiftb0y — APPROVED** [review]
Fixes confirmed to be working. Merge?

**Comment by ronso0** [comment]
> Fixes confirmed to be working. Merge?

@uklotzde?

**Comment by Holzhaus** [comment]
Uwe already approved this PR. I'll hit merge.

**Comment by Swiftb0y** [comment]
> Uwe already approved this PR.

That was before 185e373f got committed.
